File: frontend/app/api/generate.ts
import { requestOptions } from "@/lib/utils";
import { GistResponse, SummaryResponse, ChapterResponse, HighlightResponse, CustomResponse } from "@/types/generate";

const GENERATE_API_URL = `${process.env.NEXT_PUBLIC_API_URL}/generate`;

export const generateGist = async (video_id: string): Promise<GistResponse> => {
    const data = { video_id };
    try {
        const response = await fetch(`${GENERATE_API_URL}/gist`, requestOptions(data));
        return await response.json();
    } catch (error) {
        console.error('Error while generating gist:', error);
        throw error;
    }
};

export const generateSummary = async (video_id: string, prompt: string = ""): Promise<SummaryResponse> => {
    const data = { video_id, prompt };
    try {
        const response = await fetch(`${GENERATE_API_URL}/summary`, requestOptions(data));
        return await response.json();
    } catch (error) {
        console.error('Error while generating summary:', error);
        throw error;
    }
};

export const generateChapter = async (video_id: string, prompt: string = ""): Promise<ChapterResponse> => {
    const data = { video_id, prompt };
    try {
        const response = await fetch(`${GENERATE_API_URL}/chapter`, requestOptions(data));
        return await response.json();
    } catch (error) {
        console.error('Error while generating chapter:', error);
        throw error;
    }
};

export const generateHighlight = async (video_id: string, prompt: string = ""): Promise<HighlightResponse> => {
    const data = { video_id, prompt };
    try {
        const response = await fetch(`${GENERATE_API_URL}/highlight`, requestOptions(data));
        return await response.json();
    } catch (error) {
        console.error('Error while generating highlight:', error);
        throw error;
    }
};

export const generateCustom = async (video_id: string, prompt: string): Promise<CustomResponse> => {
    const data = { video_id, prompt };
    try {
        const response = await fetch(`${GENERATE_API_URL}/custom`, requestOptions(data));
        return await response.json();
    } catch (error) {
        console.error('Error while generating custom content:', error);
        throw error;
    }
};

File: backend/routes/generate.py
from fastapi import APIRouter, HTTPException
from service import TwelveLabsService
from models.generate_model import GenerateRequest, GistResponse, SummaryResponse, ChapterResponse, HighlightResponse, CustomResponse

router = APIRouter()
service = TwelveLabsService()

@router.post("/gist")
async def generate_gist_content(request: GenerateRequest) -> GistResponse:
    try:
        return service.generate_gist(request.video_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/summary")
async def generate_summary_content(request: GenerateRequest) -> SummaryResponse:
    try:
        return service.generate_summary(request.video_id, request.prompt)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/chapter")
async def generate_chapter_content(request: GenerateRequest) -> ChapterResponse:
    try:
        return service.generate_chapter(request.video_id, request.prompt)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/highlight")
async def generate_highlight_content(request: GenerateRequest) -> HighlightResponse:
    try:
        return service.generate_highlight(request.video_id, request.prompt)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/custom")
async def generate_custom_content(request: GenerateRequest) -> CustomResponse:
    try:
        if not request.prompt:
            raise HTTPException(status_code=400, detail="Prompt is required for content generation.")
        return service.generate(request.video_id, request.prompt)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
